Automattic/woocommerce-payments
# Add progressive onboarding form validation

Fixes #5474

#### Changes proposed in this Pull Request
- Install `@testing-library/react-hooks` to make React hooks testing easier.
- Implement a reusable `Field` component, to wrap `TextControl` and `CustomSelectControl`, and provide error state support.
- Create `OnboardingTextField` and `OnboardingSelectField` to use `Field` while connecting them to `OnboardingContext`. Setting up label, value, change, and validation by `name`. Allowing to render fields by just choosing the onboarding field name, like: `<OnboardingTextField name="email" />`.
- Create `OnboardingForm` component to provide submit on `enter`, validation before submit, and `nextStep()` call.
- Add `errors` and `touched` states to `OnboardingContext` to be able to track modified fields and their errors. Allowing us to show errors only after a field has been edited or the form submitted.
- Create `useValidation` hook to handle field touched and errors on change, and return `validate()` and `error()` for field components. The error message comes from `strings` and there is a generic one in case there is no matching error.
- Basic validation logic, considering our current fields I think we don't need any third party library for validation. Making all fields required and checking email seems enough.
- Update every step to use the new simplified implementation.

#### Question
Should we use a custom phone field? We already use it for WooPay. In the design it looks like a regular text field. But I wonder if we need to get it in a specific format, like separated country code. As a reference, looks like Stripe Accounts API accepts a string with no additional validation.

#### Testing instructions
>**Note**: As we're only testing validation, you can prevent redirection to Stripe by adding the following line to [handleComplete](https://github.com/Automattic/woocommerce-payments/blob/9003e42592b7fbd1669b2261858fd6c2cacd2aa0/client/onboarding-prototype/index.tsx#L36).
```js
return alert( JSON.stringify( fromDotNotation( data ), null, 2 ) );
```

- Go to `/wp-admin/admin.php?page=wc-admin&path=/payments/onboarding-prototype`.
- Fill the first step fields using an invalid email (email without @).
- The email field should be in error state – red border and error message below.
- Fix the error, and click on **Continue**.
- Click on **Continue** in the second step.
- Errors should be shown for every field.
- Continue until the full flow is done.
- The alert should contain the provided data – If you opted to use the mentioned snippet. Otherwise, you can check it by using breakpoints or looking at prefilled details on Stripe.

-------------------


**Post merge**
